### PR TITLE
added simple initializer to some modules

### PR DESCRIPTION
**Pull request by moskomule**
Hi, I thought `nn.init` is great but initializing variables in layers using `nn.init` is a bit complicated. So I try to implement easier way of initialization like below. 

```python
m = nn.Linear(20, 30, initializer=lambda x: init.xavier_normal(x, 1))
```

I do not think this is the best way and thus implemented only for `Linear`. If you think this is useful, give me advice and help me. Thank you.

### CODE REVIEW

**Comment by Kaixhin** [comment]
The method you suggested imposes the same initialisation on weights and biases, and more generally, might be ambiguous with parameters in other layers (like the mean of batch normalisation).

I would suggest [chaining](https://en.wikipedia.org/wiki/Method_chaining), along the lines of the original [`nninit`](https://github.com/Kaixhin/nninit#example).

**Comment by apaszke** [comment]
I agree that this is not an ideal solution, as it's quite ambiguous. Chaining is a bit annoying in Python, because it has to happen inside parenthesis (otherwise it's a syntax error because the whitespace has meaning).

**Comment by Kaixhin** [comment]
@apaszke what do you mean exactly? I was thinking along the lines of:

```py
m = nn.Linear(20, 30).init('weight', nn.init.xavier_normal, 1)

# Linear.init pseudocode:
def init(self, accessor, method, **kwargs):
    method(self[accessor], **kwargs)
    return self
```

First argument [accesses the field](https://github.com/Kaixhin/nninit#accessors), second [specifies the method](https://github.com/Kaixhin/nninit#initialisers), rest are passed into the method. I still haven't used Python that much, but is there a problem with the above?

**Comment by apaszke** [comment]
the above is fine, however this is a syntax error:
```python
nn.Linear(20, 30).init('weight', ...)
                 .init('bias', ...)
```
You need extra parenthesis to fix it:
```python
m = (nn.Linear(20, 30).init('weight', ...)
                      .init('bias', ...))
```

**Comment by apaszke** [comment]
But I don't have any better ideas, so I guess we could add that

**Comment by Kaixhin** [comment]
Ah I see what you mean now. Yeah that is a bit annoying, but having longer lines or introducing an extra parenthesis isn't the end of the world. I'm obviously biased, but I think chaining is a neat way to approach this. Sure was an improvement on [my first API for `nninit`](https://github.com/Kaixhin/nninit/tree/ac110397f4ac6827413c60a9c02da9af7143bc19#example).

**Comment by moskomule** [comment]
Thank you for your ideas. Yes, as @Kaixhin mentions my original idea did not consider bias well. 
And @apaszke 's 

```python
m = (nn.Linear(20, 30).init('weight', ...).init('bias', ...))
```

is nice, but I don't think this is Python's way. What do you think of using dictionary instead, for example 

```python
m = nn.Linear(20, 30, initializer={'weight': ..., 'bias': ...})
```

?

**Comment by Kaixhin** [comment]
Yep that seems more Pythonic but has the right level of flexibility 👍

Now this can be added into [`nn.Module`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py) directly so it can be used with convolutions, RNNs, everything. Loop over the keys and throw an error if the module doesn't contain it.

**Comment by moskomule** [comment]
Which do you think is better, adding `reset_paramters` to `nn.Module` directly or modifying `reset_parameters` like I pushed?

File: torch/nn/modules/conv.py [hunk 1]
```diff
@@ -10,7 +10,8 @@
 class _ConvNd(Module):
 
     def __init__(self, in_channels, out_channels, kernel_size, stride,
-                 padding, dilation, transposed, output_padding, groups, bias):
+                 padding, dilation, transposed, output_padding, groups, bias,
+                 initializer=dict()):
```
Comment (fmassa) [inline torch/nn/modules/conv.py]: It's not good to have default values of [functions to be mutable elements](http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/). Might be better to set it to `None` and create the dict inside the function
```python
def fun(initializer=None):
    if initializer is None:
        initializer = dict()
```

Comment (moskomule) [inline torch/nn/modules/conv.py]: OK, thank you.

File: torch/nn/modules/conv.py [hunk 2]
```diff
@@ -35,16 +36,31 @@ def __init__(self, in_channels, out_channels, kernel_size, stride,
...
     def reset_parameters(self):
         n = self.in_channels
         for k in self.kernel_size:
             n *= k
         stdv = 1. / math.sqrt(n)
-        self.weight.data.uniform_(-stdv, stdv)
-        if self.bias is not None:
-            self.bias.data.uniform_(-stdv, stdv)
+
+        if isinstance(self.initializer, function):
```
Comment (fmassa) [inline torch/nn/modules/conv.py]: This looks a bit verbose. You can probably simplify it by using the right initializers when no default argument was passed?

File: torch/nn/modules/linear.py [hunk 3]
```diff
@@ -38,32 +38,26 @@ def __init__(self, in_features, out_features, bias=True, initializer=None):
...
-            self.weight.data.uniform_(-stdv, stdv)
-        else:
-            weight_initializer(self.weight)
-        if self.bias is not None:
-            if bias_initializer is None:
-                self.bias.data.uniform_(-stdv, stdv)
-            else:
-                bias_initializer(self.bias)
+        weight_initializer(self.weight)
+        if self.bias is None:
```
Comment (fmassa) [inline torch/nn/modules/linear.py]: maybe this should be instead
```python
if self.bias is not None:
```

**Comment by moskomule** [comment]
Now these layers seem not be able to pickle but I don't know how to fix...

**Comment by fmassa** [comment]
@moskomule I think you can't pickle `lambda` functions. Do you need to add `initializers` to `self`?

**Comment by moskomule** [comment]
While Python 3.6's test has passed, others haven't. Do you have any idea to solve it?

**Comment by Kaixhin** [comment]
@moskomule that's not you, sorry - looks like a test broke on some other versions. Pasting stack trace here temporarily so others don't need to dive in:

```sh
======================================================================
FAIL: test_AdaptiveMaxPool2d_tuple (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 2852, in <lambda>
    setattr(TestNN, test_name, lambda self, test=test: test(self))
  File "/home/travis/build/pytorch/pytorch/test/common_nn.py", line 571, in __call__
    self._do_test(test_case, module, input)
  File "test_nn.py", line 82, in _do_test
    test_case.check_jacobian(module, input, self.jacobian_input)
  File "/home/travis/build/pytorch/pytorch/test/common_nn.py", line 459, in check_jacobian
    PRECISION
AssertionError: 0.10383865983731866 not less than or equal to 1e-05
```

**Comment by moskomule** [comment]
@Kaixhin Python 2.7's test is again Pickling Error as below, so I though it is because of my pull request. 

```
======================================================================
ERROR: test_serialization_container (__main__.TestTorch)
----------------------------------------------------------------------
Traceback (most recent call last):

  File "test_torch.py", line 3507, in test_serialization_container
    torch.save(module.Net(), checkpoint)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/torch/serialization.py", line 120, in save
    return _save(obj, f, pickle_module, pickle_protocol)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/torch/serialization.py", line 186, in _save
    pickler.dump(obj)
PicklingError: Can't pickle <type 'instancemethod'>: attribute lookup __builtin__.instancemethod failed
```

**Comment by moskomule** [comment]
Do you have any idea to solve Pickling Error in Python2.7 below? I don't use Python 2.7 so I'm not sure the reason.

```
PicklingError: Can't pickle <type 'instancemethod'>: attribute lookup __builtin__.instancemethod failed
```

**Comment by apaszke** [comment]
Lambdas are not serializable. If you want to keep them as attributes, you need to remove them in `__getstate__`

**Comment by moskomule** [comment]
I'd removed all lambda expression except in doctest..

**Comment by soumith** [comment]
this PR is failing python 2.7. Here's the stacktrace:

```
======================================================================
ERROR: test_serialization_container (__main__.TestTorch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_torch.py", line 3687, in test_serialization_container
    torch.save(module.Net(), checkpoint)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/torch/serialization.py", line 120, in save
    return _save(obj, f, pickle_module, pickle_protocol)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/torch/serialization.py", line 186, in _save
    pickler.dump(obj)
PicklingError: Can't pickle <type 'instancemethod'>: attribute lookup __builtin__.instancemethod failed

```

**Comment by moskomule** [comment]
@soumith Yes, but I am not certain what causes this failing. Now I removed lambda expressions in docstring.

File: torch/nn/modules/conv.py [hunk 4]
```diff
@@ -10,7 +11,8 @@
 class _ConvNd(Module):
 
     def __init__(self, in_channels, out_channels, kernel_size, stride,
-                 padding, dilation, transposed, output_padding, groups, bias):
+                 padding, dilation, transposed, output_padding, groups, bias,
+                 initializer):
```
Comment (alykhantejani) [inline torch/nn/modules/conv.py]: we could provide a default value here of `initializer = {}`, then in the constructor we can just do `self.initializer = initializer` and in `reset_parameters` we can take advantage of the default response of the`get` method on dicts. i.e. `reset_parameters` would be something like:

```python
self.initializer.get("weight", self._initializer)(self.weight)
if bias is not None:
    self.initializer.get("bias", self._initializer(self.bias)
```

File: torch/nn/modules/conv.py [hunk 5]
```diff
@@ -109,6 +123,8 @@ class Conv1d(_ConvNd):
         dilation (int or tuple, optional): Spacing between kernel elements
         groups (int, optional): Number of blocked connections from input channels to output channels
         bias (bool, optional): If True, adds a learnable bias to the output
+        initializer(dict, optional): dictionary of initializer of weights and bias, if None (default),
```
Comment (alykhantejani) [inline torch/nn/modules/conv.py]: probably worth mentioning the dict needs keys `weight` and `bias`

File: torch/nn/modules/conv.py (hunk 4 above)
Comment (alykhantejani) [inline torch/nn/modules/conv.py]: Sorry, this default can be passed in, in the actual implementations i.e. `Conv1d` etc

File: torch/nn/modules/conv.py (hunk 5 above)
Comment (moskomule) [inline torch/nn/modules/conv.py]: Thank you for advice, is `dictionary of initializer of weights and bias (for instance {"weight": lambda x: init.kaiming_uniform(x)}), if None (default), they are uniformly initialized.` enough?

**Comment by moskomule** [comment]
Is this PR still alive, I mean, do I need to do something?

**Pull request closed without merging**